nwb_conversion_tools/utils/basenwbephyswriter.py

```diff
@@ -412,43 +420,88 @@ def add_electrical_series(self, segment_index):
-            channel_offset = self._get_channel_property_values("offset")
-        unsigned_coercion = channel_offset / channel_conversion
-        if not np.all([x.is_integer() for x in unsigned_coercion]):
-            raise NotImplementedError(
-                "Unable to coerce underlying unsigned data type to signed type, which is currently required for NWB "
-                "Schema v2.2.5! Please specify 'write_scaled=True'."
-            )
-        elif np.any(unsigned_coercion != 0):
-            warnings.warn(
-                "NWB Schema v2.2.5 does not officially support channel offsets. The data will be converted to a signed "
-                "type that does not use offsets."
-            )
-            unsigned_coercion = unsigned_coercion.astype(int)
+        unsigned_coercion = None
+        # this works for both SI013 and SI090
+        channel_gains = np.ones(len(self._get_channel_ids()), dtype="int")
+        channel_offsets = np.zeros(len(self._get_channel_ids()), dtype="int")
+        data_chunk_dtype = None
+
+        # if writing scaled traces, we don't need to store gain and offset as columns
+        if not self._conversion_ops["write_scaled"]:
+            if self._get_gains() is not None and self._get_offsets() is not None:
+                channel_gains = self._get_gains()
+                channel_offsets = self._get_offsets()
+                recording_dtype = np.dtype(self._get_dtype(self._conversion_ops["write_scaled"]))
+                if np.issubdtype(recording_dtype, np.unsignedinteger):
+                    # only when dtype is unsigned we use the unsigned_coercion mechanism
+                    unsigned_coercion = channel_offsets / channel_gains
+                    if not np.all([x.is_integer() for x in unsigned_coercion]):
+                        raise NotImplementedError(
+                            "Unable to coerce underlying unsigned data type to signed type, which is currently "
+                            "required for NWB Schema v2.2.5! Please specify 'write_scaled=True'."
+                        )
+                    elif np.any(unsigned_coercion != 0):
+                        warnings.warn(
+                            "NWB Schema v2.2.5 does not officially support channel offsets. The data will be "
+                            "converted to a signed type that does not use offsets."
+                        )
+                        unsigned_coercion = unsigned_coercion.astype(int)
+                    unsigned_coercion = list(unsigned_coercion)
+
+                    # in this case we have to force the uint type to int
+                    dtype_str = recording_dtype.str
+                    # to avoid overflow, we need to double the number of bytes
+                    nbits = int(dtype_str[dtype_str.find("u") + 1 :])
+                    new_dtype_str = f"{dtype_str[0]}i{2*nbits}"
+                    data_chunk_dtype = np.dtype(new_dtype_str)
+
```

Comment:
This seems like a lot of untested internal logic/machinery for handling `dtype` promotion; I'll investigate established ways of doing it.